### PR TITLE
TLOZ: Fix non-deterministic item pool generation

### LINKED ISSUE #3664: Bug: Non-deterministic generation when combining SMZ3 and The Legend of Zelda
### What happened?

Generating a multiworld with a fixed seed containing both an SMZ3 world and a The Legend of Zelda world results in non-deterministic generation. Sometimes multiple generations will produce very similar results, but other times the results are very different.

Generating either world on its own appears to be deterministic, it is only when both are together. I could not get the same non-deterministic behaviour to occur with other combinations of games with either of these two worlds (including 2x SMZ3 or 2x The Legend of Zelda), only both together.

I've tested on v0.5.0 (both from source and from the Windows installation) and on what was the HEAD of the `main` branch at the time (925e02dca72420d94971db28c2c15c4ddd3092c2)

For each generation, I generated with template SMZ3 and The Legend of Zelda yamls, but with a meta.yaml used to disable progression balancing, they are attached as Players.zip.
[Players.zip](https://github.com/user-attachments/files/16289822/Players.zip)

Attached are the spoiler logs for two generations ran from v0.5.0 on source with `python -O Generate.py --seed 1`.
I'm not too familiar with what is important in these games, however, looking at the playthrough, in gen1's Sphere 2, it picks up `Missile (green Brinstar pipe) (Player1): Red Candle (Player2)`, but in gen2 (also in Sphere 2), this is , `Missile (green Brinstar pipe) (Player1): Heart Container (Player2)` instead, and if it's in the playthrough, then I'm expecting it to be progression.
[gen1_AP_14089154938208861744_Spoiler.txt](https://github.com/user-attachments/files/16289842/gen1_AP_14089154938208861744_Spoiler.txt)
[gen2_AP_14089154938208861744_Spoiler.txt](https://github.com/user-attachments/files/16289843/gen2_AP_14089154938208861744_Spoiler.txt)

Since I can only get this non-deterministic behaviour to occur when both worlds are present, I'm not sure where to even begin looking for the cause of the issue in these worlds, so I've made this report.

### What were the expected results?

Given the same seed, generation should be the same each time (progression items at least, I don't know if filler is also supposed to be deterministic).

### Software

Local generation

### PR DESCRIPTION
**Pull request by Mysteryem**
## What is this fixing or adding?

The way the item pool was constructed involved iterating unions of sets. Sets are unordered, so the order of iteration of these combined sets would be non-deterministic, resulting in the items in the item pool being generated in a different order with the same seed.

Rather than creating unions of sets at all, the original code has been replaced with using Counter objects. As a dict subclass, Counter maintains insertion order, and its update() method makes it simple to combine the separate item dictionaries into a single dictionary with the total count of each item across each of the separate item dictionaries.

Fixes #3664 - After investigating more deeply, the only differences I could find between generations of the same seed was the order of items created by TLOZ. The issue appears to have occurred when generating with SMZ3 because SMZ3 removes junk/trap items from the pool during pre_fill, specifically in its `JunkFillGT()` and it could pick different items with the same seed because TLOZ's items were in a non-deterministic order, so this patch appears to fix the non-deterministic generation issue.

## How was this tested?

I modified Main.py to write out the locations of all progression items and what those items were into a file on a single line, but only if the line did not already exist in the file, this way, I ran many generations using `--seed 1 --skip_output` and if there was more than one line in the file at the end, then there was non-deterministic generation.

The simple case of just the template TLOZ and SMZ3 yamls that commonly resulted in non-deterministic generation was tested, as well as with many other template yamls that are known to produce deterministic results..

### CODE REVIEW

**Review by t3hf1gm3nt — APPROVED** [review]
LGTM. I appreciate you digging deeper into this Mysteryem.

File: worlds/tloz/ItemPool.py [hunk 1]
```diff
@@ -2,6 +2,8 @@
 from .Locations import level_locations, all_level_locations, standard_level_locations, shop_locations
 from .Options import TriforceLocations, StartingPosition
 
+from collections import Counter
```
Comment (Exempt-Medic) [inline worlds/tloz/ItemPool.py:5]: Technically, this import should be at the top

Comment (Mysteryem) [reply worlds/tloz/ItemPool.py:5]: I don't see a mention of import order in the style guide. Is the desired order mentioned somewhere?

Comment (Exempt-Medic) [reply worlds/tloz/ItemPool.py:5]: https://peps.python.org/pep-0008/#imports

>Imports should be grouped in the following order:
> 1. Standard library imports.
> 1. Related third party imports.
> 1. Local application/library specific imports.
> 
> You should put a blank line between each group of imports.

**Review by Exempt-Medic — APPROVED** [review]
The change is simple enough, tested some configurations to find a non-deterministic generation which, as far as my testing can tell, is deterministic with the change

**Comment by Mysteryem** [comment]
I've updated the description because it looks like the issue occurred specifically with SMZ3 because it removes junk/trap items from the multiworld's item pool during pre_fill and then places them in GT. Location filling code in core, e.g. distribute_items_restrictive, appears to defend against items/locations being created in a non-deterministic order by sorting the item/location pools before shuffling them, but because of SMZ3, different items could have been removed from the item pool each time.

I think the one time I managed to get TLOZ to produce non-deterministic results on its own was probably a mistake on my part, likely from uncommitted changes laying around from a different branch I had been working on.

**Review by NewSoupVi — APPROVED** [review]

**Pull request merged**